src/web/middleware.rs

```diff
@@ -131,6 +131,50 @@ impl Middleware<ServerState> for DbTransaction {
     }
 }
 
+#[derive(Debug)]
+pub struct PreConditionCheck;
+
+impl Middleware<ServerState> for PreConditionCheck {
+    /// This middleware must be wrapped by the `DbTransaction` middleware to ensure a Db object
+    /// is available.
+    fn start(&self, req: &HttpRequest<ServerState>) -> Result<Started> {
+        let precondition =
+            match <Option<PreConditionHeader> as FromRequest<ServerState>>::from_request(&req, &())
+            {
+                Ok(Some(precondition)) => precondition,
+                Ok(None) | Err(_) => return Ok(Started::Done),
+            };
+        let user_id = HawkIdentifier::from_request(&req, &())?;
+        let db = <Box<dyn Db>>::from_request(&req, &())?;
+        let collection = CollectionParam::from_request(&req, &())
+            .ok()
+            .map(|v| v.collection);
+        let bso = BsoParam::from_request(&req, &()).ok().map(|v| v.bso);
+        let fut = db
+            .extract_resource(user_id, collection, bso)
+            .and_then(move |resource_ts| {
+                let (status, modified) = match precondition {
+                    PreConditionHeader::IfModifiedSince(header_ts) if resource_ts <= header_ts => {
+                        (StatusCode::NOT_MODIFIED, header_ts)
```

Comment:
Unless I'm mistaken python's rendering resource_ts as the X-Last-Modified in these cases, not header_ts?

Reply:
This is purely the status code for the request, not the header.

Reply:
I meant the second item of these tuple set as X-Last-Modified below

Reply:
Ah! Right.